csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3620 @@
+  // replays will modify the loop map, we operate on a copy of the loop map,
+  // not the original one.
+
+  auto loop_graph_copy = idGraph(IdMappingMode::LOOP);
+  std::unordered_map<IdGroup, IterDomain*> loop_graph_copy_promotion_map;
+
+  for (auto loop_group : loop_graph_copy.disjointIdSets().disjointSets()) {
+    if (loop_group->size() == 1) {
+      loop_graph_copy_promotion_map[loop_group] = loop_group->front();
+      continue;
+    }
+
+    // We need to check the exact groups the terminal id's are in, but for
+    // promotion we want an iter domain within the loop group. Since exact
+    // group can traverse loop group boundaires, save a vector of the group
+    // and the iter domain.
+    std::vector<std::pair<IdGroup, IterDomain*>> exact_promoted_terminal_ids;
+    for (auto loop_id : *loop_group) {
+      if (terminal_loop_ids.has(loop_id)) {
+        auto iel_set_pair =
+            intersection_exact_loop_graph.disjointIdSet(loop_id);
+        TORCH_INTERNAL_ASSERT(iel_set_pair.second);
+        auto iel_group = iel_set_pair.first;
+        auto iel_promo_it = iel_promotion_map.find(iel_group);
+        if (iel_promo_it == iel_promotion_map.end()) {
+          auto promo_id_exact_it =
+              idGraph(IdMappingMode::EXACT).disjointIdSet(loop_id);
+          TORCH_INTERNAL_ASSERT(promo_id_exact_it.second);
+          exact_promoted_terminal_ids.push_back(
+              std::make_pair(promo_id_exact_it.first, loop_id));
+        } else {
+          auto promo_id_exact_it =
+              idGraph(IdMappingMode::EXACT).disjointIdSet(iel_promo_it->second);
+          TORCH_INTERNAL_ASSERT(promo_id_exact_it.second);
+          exact_promoted_terminal_ids.push_back(
+              std::make_pair(promo_id_exact_it.first, iel_promo_it->second));
+        }
+      }
+    }
+
+    // All exact groups with iter domains in this loop group
+    IdGroups exact_groups;
+    for (auto loop_id : *loop_group) {
+      auto exact_set_pair =
+          idGraph(IdMappingMode::EXACT).disjointIdSet(loop_id);
+      TORCH_INTERNAL_ASSERT(exact_set_pair.second);
+      exact_groups.pushBack(exact_set_pair.first);
+    }
+
+    // All exact groups covered by all iter domains in this loop group
+    IdGroups loop_group_covered_ids;
+    for (auto exact_group : exact_groups) {
+      auto covered_it = exact_covered_ids.find(exact_group);
+      TORCH_INTERNAL_ASSERT(covered_it != exact_covered_ids.end());
+      loop_group_covered_ids.pushBack(covered_it->second);
+    }
+
+    IterDomain* loop_promotion_id = nullptr;
+
+    for (auto entry : exact_promoted_terminal_ids) {
+      auto terminal_id_group = entry.first;
+      auto terminal_id = entry.second;
+      auto covered_it = exact_covered_ids.find(terminal_id_group);
+      TORCH_INTERNAL_ASSERT(covered_it != exact_covered_ids.end());
+      if (loop_group_covered_ids.subtract(covered_it->second).size() == 0) {
+        loop_promotion_id = terminal_id;
```

Comment:
Looks like there can be multiple `terminal_id_group` that hits here. Since `exact_promoted_terminal_ids` are deterministically ordered, this part should also be deterministic.

Reply:
Yes. I've tried to keep everything deterministic, but if you see something that doesn't look like it, let me know.